sw/host/opentitanlib/src/debug/dmi.rs

```diff
@@ -0,0 +1,332 @@
+    pub const DATA0: u32 = 0x04;
+    pub const DATA1: u32 = 0x05;
+    pub const DMCONTROL: u32 = 0x10;
+    pub const DMSTATUS: u32 = 0x11;
+    pub const HARTINFO: u32 = 0x12;
+    pub const ABSTRACTCS: u32 = 0x16;
+
+    pub const DMSTATUS_ANYHALTED_MASK: u32 = 1 << 8;
+    pub const DMSTATUS_ANYRUNNING_MASK: u32 = 1 << 10;
+    pub const DMSTATUS_ANYUNAVAIL_MASK: u32 = 1 << 12;
+    pub const DMSTATUS_ANYNONEXISTENT_MASK: u32 = 1 << 14;
+    pub const DMSTATUS_ANYRESUMEACK_MASK: u32 = 1 << 16;
+    pub const DMSTATUS_ANYHAVERESET_MASK: u32 = 1 << 18;
+
+    pub const DMCONTROL_DMACTIVE_MASK: u32 = 1 << 0;
+    pub const DMCONTROL_NDMRESET_MASK: u32 = 1 << 1;
+    pub const DMCONTROL_HASEL_MASK: u32 = 1 << 26;
+    pub const DMCONTROL_ACKHAVERESET_MASK: u32 = 1 << 28;
+    pub const DMCONTROL_RESUMEREQ_MASK: u32 = 1 << 30;
+    pub const DMCONTROL_HALTREQ_MASK: u32 = 1 << 31;
+
+    pub const DMCONTROL_HARTSELHI_SHIFT: u32 = 6;
+    pub const DMCONTROL_HARTSELLO_SHIFT: u32 = 16;
+
+    pub const ABSTRACTCS_CMDERR_MASK: u32 = (1 << 11) - (1 << 8);
+    pub const ABSTRACTCS_BUSY_MASK: u32 = 1 << 12;
+
+    pub const ABSTRACTCS_CMDERR_SHIFT: u32 = 8;
+
+    pub const ABSTRACTCS_CMDERR_NONE: u32 = 0;
+}
+
+use consts::*;
+
+/// Debug module interface (DMI) abstraction.
+pub trait Dmi {
+    /// Read a DMI register.
+    fn dmi_read(&mut self, addr: u32) -> Result<u32>;
+
+    /// Write a DMI register.
+    fn dmi_write(&mut self, addr: u32, data: u32) -> Result<()>;
+}
+
+impl<T: Dmi> Dmi for &mut T {
+    fn dmi_read(&mut self, addr: u32) -> Result<u32> {
+        T::dmi_read(self, addr)
+    }
+
+    fn dmi_write(&mut self, addr: u32, data: u32) -> Result<()> {
+        T::dmi_write(self, addr, data)
+    }
+}
+
+/// DMI interface via OpenOCD.
+pub struct OpenOcdDmi {
+    openocd: OpenOcd,
+    tap: String,
+    abits: u32,
+}
+
+impl OpenOcdDmi {
+    /// Create a new DMI interface via OpenOCD.
+    ///
+    /// This should be an OpenOCD instance with JTAG scan chain already set up,
+    /// but not with target set up. If target has been set up, OpenOCD will access
+    /// DMI registers on its own, which will interfere with raw DMI operations.
+    pub fn new(mut openocd: OpenOcd, tap: &str) -> Result<Self> {
+        openocd.irscan(tap, DTMCS)?;
+        let res = openocd.drscan(tap, 32, DTMCS_DMIRESET_MASK)?;
+        let abits = (res & DTMCS_ABITS_MASK) >> DTMCS_ABITS_SHIFT;
+
+        openocd.irscan(tap, DMI)?;
+        Ok(Self {
+            openocd,
+            tap: tap.to_owned(),
+            abits,
+        })
+    }
+
+    fn dmi_op(&mut self, op: u64) -> Result<u64> {
+        let res = self
+            .openocd
+            .drscan(&self.tap, self.abits + DMI_ADDRESS_SHIFT, op)?;
+
+        // We just scanned into the DMI register, so the scanned result should be empty.
+        ensure!(res == 0, "Unexpected DMI initial response {res:#x}");
+
+        // Run the DMI operation.
+        self.openocd.execute("runtest 10")?;
```

Comment:
This function assumes that 10 cycles will suffice to execute the operation, which may not be the case for all operations? Should this duration be at least configurable/documented? Alternatively the function could retry the operation with a longer timeout if it does not succeed.

Reply:
Documented. I'd like to avoid the complexity of adding a retry logic